### DISCUSSION TITLE
Custom Jinja2 Filters

### DISCUSSION THREAD

**Discussion by gmazoyer** [peering-manager/peering-manager#383] [Ideas]
As part of issue #356, I'd like to bring some useful [Jinja2](https://jinja.palletsprojects.com/) filters to simplify template writing.

As of this moment, templates, depending on your needs, can be quite complex to write and can involve lots of loops, checks and data merges. To make all of this a bit more user friendly, it is possible to expose some filters (functions) that will do the complex processing for you. This is what Ansible does (for instance).

In order to understand user needs, I would like to encourage people interested in this to share what they want to see as filters. Please keep in mind that these filters must be as much as possible generic so everyone can benefit from it. If you have ideas, please use the following template to propose them.

```
Filter name (lowercases, underscores for separator): infinity_gauntlet
Description (what the filter should do): Holds all Infinity Gems to gather Universe power
Example (what should look like in templates): {{ all_gems|infinity_gauntlet }}
Input data (what data it should process): [ "mind", "power", "reality", "soul", "space", "time" ]
Output data (what result it should produce): "snap"
```

**Comment by mngan** [comment]
```
Filter name: iter_sessions
Description: Return an iterator that yields tuples of address family and session for IX or BGP Group objects,  
sorted by address family and session IP.
Example: 
{% for ix in internet_exchanges %}
  {% for address_family, session in ix | iter_sessions %} 
    ... 
  {% endfor %}
{% endfor %}
{% for group in bgp_groups %}
  {% for address_family, session in group | iter_sessions %} 
    ... 
  {% endfor %}
{% endfor %}
Input data: IX or BGP Group object/dict
Output data: [ (6, session dict1), (6, session_dict2), (4, session_dict3), ...]
```

Another possibility is to have one that can iterate over all the sessions from a list of IX's or BGP Groups. This could reduce the levels of iteration further, but could cause you to require more tests to know when the IX/Group change, if that is necessary. e.g

```
{% for ix, address_family, session in internet_exchanges | iter_all_sessions %}
  {% if prev_ix != ix %}
    ...
  {% endif %}
  ...
  {% set prev_ix = ix %}
{% endfor %}
{% for group, address_family, session in bgp_groups | iter_all_sessions %}
  {% if prev_group != group %}
    ...
  {% endif %}
  ...
  {% set prev_group = group %}
{% endfor %}
```

**Reply by mngan** [reply]
Looking at the way the configuration context is constructed now, my suggesting will probably change a little. Since these are no longer dicts, but Django querysets, then perhaps something more like:

```
{% for ix in internet_exchanges %}
  {% for ipversion in (6, 4) %}
    {% for session in ix | sessions(ipversion) %} 
    ... 
    {% endfor %}
  {% endfor %}
{% endfor %}
{% for group in bgp_groups %}
  {% for ipversion in (6, 4) %}
    {% for session in group | sessions(ipversion) %} 
    ... 
    {% endfor %}
  {% endfor %}
{% endfor %}
```

**Comment by mngan** [comment]
```
Filter Name: iter_(import|export)_policy
Description: For a session object, return the iterator (or list if that is easier) for all inherited of polices from the least to most the specific (e.g BGP Group/IX,  ASN, Session) for import and in the reverse order for export. 
Example: 
import [ {{session | iter_import_policy | map(attribute='slug') | join(' ')}} ]
export [ {{session | iter_export_policy | map(attribute='slug') | join(' ')}} ]
Input Data: session object with relationships to IX, BGP Group and ASN
Output Data: iterator/list of policy objects [ {name:'Policy A', slug:'policy-a', ...}, {name:'Policy B', slug='policy-b'} ]
```

Just wondering if is will be necessary to de-duplicate the list of policies in case they are specified in multiple places. In that case, who wins? If this were evaluated on JUNOS without de-duplication, the first instance will always win, but unsure how this would behave on other platforms.

**Comment by mngan** [comment]
I use the contact fields in descriptions, but these need sanitizing to work properly. Not sure if there is a better way to do this that works on all platforms. Mainly, ensure all the data coming from the contact fields is ASCII (not sure how many configs support unicode) and strip any quotes from the string that could confuse the config string format. 

```
Filter Name: safe_string
Description: Make strings safe(r) to be put into descriptions/comments
Example:
{%- macro session_description(session) -%}"{{ session.autonomous_system.name | safe_string }}{% if session.autonomous_system.contact_email %} - {{ session.autonomous_system.contact_name | safe_string }} ({{session.autonomous_system.contact_email | safe_string }}){% endif %}"{%- endmacro -%}
Input Data: Jörgen "Kef" Lindman
Output Data: Jorgen Kef Lindman
```

I currently have something like this, which mostly works, but could be better. 

```
{%- macro session_description(session) -%}"{{ session.autonomous_system.name.encode('ascii', 'xmlcharrefreplace').decode() | replace('\"', '') }}{% if session.autonomous_system.contact_email %} - {{ session.autonomous_system.contact_name.encode('ascii', 'xmlcharrefreplace').decode() | replace('\"', '') }} ({{session.autonomous_system.contact_email.encode('ascii', 'xmlcharrefreplace').decode() | replace('\"', '') }}){% endif %}"{%- endmacro -%}
```

**Comment by mngan** [comment]
Since the context is now mostly querysets and not regular list/dict, then some assistance here might be useful to extra/manipulate the queryset. This will abstract away from the representation you are actually using. In some ways, you end up replicating a bunch of standard Jinja filters for list/dict types, but it will allow you to change the underlying representations more easily later without disrupting users too much. 

Common operations:
- `length`
- `sort`/`dictsort`
- `select`/`selectattr`

**Comment by mngan** [comment]
A filter to pull out the route servers sessions for an IX would be useful

I have code like this:

```
 {%- set rs = internet_exchange.sessions.values() | sum(start=[]) | selectattr("is_route_server") | list %}
  {%- if rs | length > 0 %}
        term route-server {
          from {
            neighbor [ {{ rs | map(attribute="ip_address") | map('string') | sort | join(' ') }} ];
          }
          then metric add 10; 
        }
  {%- endif %}
```

Which could be be more simple like this:
```
 {%- set rs = internet_exchange |  route_server('ip_address') %}
  {%- if rs | length > 0 %}
        term route-server {
          from {
            neighbor [ {{ rs | map('string') | sort | join(' ') }} ];
          }
          then metric add 10; 
        }
  {%- endif %}
```

**Comment by mngan** [comment]
A filter to return a unique list of AutonomousSystem objects for an IX/BGPGroup could also be useful. Right now I construct duplicate policies because I don't have a good way to get the list per IX or BGPGroup. There is the global list, but I would like it to be a little more granular since the policies generated from these lists could be different. 

Or possibly, invert the query? Given the list of ASN objects, return the ones that have IX sessions on this router? or Direct Sessions on this router? 

```
{%- for asn in autonomous_systems | ix_peers %}
...
{% endfor %}
{%- for asn in autonomous_systems | ix_peers(ix_name) %}
...
{% endfor %}
{%- for asn in autonomous_systems | direct_peers %}
...
{% endfor %}
{%- for asn in autonomous_systems  | direct_peers(group_name) %}
...
{% endfor %}
```

**Comment by gmazoyer** [comment]
Some work has been done in PR #391 that address quite a bit of your real interesting suggestions. If you guys have some spare time and want to take a look, I'll wait for your feedbacks before merging the PR.

**Reply by mngan** [reply]
We don't currently use this feature, but I was just looking at the "missing sessions" email template. I don't think that there is a `missing_sessions` method for an InternetExchange. So this probably needs to be another filter that you can also specify an ASN to filter the missing sessions for?

**Reply by mngan** [reply]
Also related to email templates, with the new connection model, it isn't so easy to pull the IP addresses for an IX since there could be multiple connections. Having a filter to retrieve the IX local IP's would be useful.

**Reply by mngan** [reply]
I think there is a bug with the `route_servers` filter is missing the address_family option. Docs says you can pass in the address family, but the filter code doesn't accept an argument.

**Reply by jamesditrapani** [reply]
@mngan Cheers. Have opened #412 to track your feedback.
